### PR TITLE
Adds support for HostedZoneName instead of HostedZoneId in Domain section of the Api

### PR DESCRIPTION
**Pull request by mbarneyjr**
*Issue #, if available:*

Issue #1402 

*Description of changes:*

Adds support for inclusion of `HostedZoneName` instead of `HostedZoneId` in the `Domain` section of an `AWS::Serverless::Api`. This allows users to not need to worry about looking up the hosted zone ID, and letting CloudFormation handle it for them via the `AWS::Route53::RecordSetGroup` resource.

`HostedZoneName` is forwarded to the `RecordSetGroup` resource just as `HostedZoneId` is. 

The following should now be valid SAM:

```yml
Globals:
  Api:
    Domain:
      DomainName: www.my-domain.com
      CertificateArn: my-valid-cert-arn
      EndpointConfiguration: EDGE
      Route53:
        HostedZoneName: my-domain.com.
```

*Description of how you validated changes:*

Not 100% sure how to validate the changes, didn't immediately see any documentation around it

*Checklist:*

- [x] Write/update tests
- [x] `make pr` passes (It doesn't have any more failures than what is on `master`)
- [x] Update documentation
- [x] Verify transformed template deploys and application functions as expected
- [x] Add/update example to `examples/2016-10-31`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: examples/2016-10-31/custom_domains_with_route53_hosted_zone_name/template.yaml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      StageName: Prod
+      Domain:
+        DomainName: !Ref DomainName
+        CertificateArn: !Ref ACMCertificateArn
+        EndpointConfiguration: EDGE
+        BasePath:
+          - /home
+        Route53:
+         HostedZoneName: www.my-domain.com.
+         IpV6: true
```
Comment (timoschilling) [inline examples/2016-10-31/custom_domains_with_route53_hosted_zone_name/template.yaml]: I think `IpV6` is also optional

File: samtranslator/model/api/api_generator.py [hunk 2]
```diff
@@ -338,15 +338,18 @@ def _construct_api_domain(self, rest_api):
         record_set_group = None
         if self.domain.get("Route53") is not None:
             route53 = self.domain.get("Route53")
-            if route53.get("HostedZoneId") is None:
+            if route53.get("HostedZoneId") is None and route53.get("HostedZoneName") is None:
                 raise InvalidResourceException(
                     self.logical_id, "HostedZoneId is required to enable Route53 support on Custom Domains."
```
Comment (praneetap) [inline samtranslator/model/api/api_generator.py]: nit: "HostedZoneId or HostedZoneName is required to enable Route53 support on Custom Domains."

Comment (timoschilling) [inline samtranslator/model/api/api_generator.py]:
```suggestion
                    self.logical_id, "HostedZoneId or HostedZoneName is required to enable Route53 support on Custom Domains."
```

File: examples/2016-10-31/custom_domains_with_route53_hosted_zone_name/template.yaml (hunk 1 above)
Comment (mbarneyjr) [reply examples/2016-10-31/custom_domains_with_route53_hosted_zone_name/template.yaml]: This is an example, not saying these things are required, right?

File: samtranslator/model/api/api_generator.py (hunk 2 above)
Comment (mbarneyjr) [reply samtranslator/model/api/api_generator.py]: Missed that

File: examples/2016-10-31/custom_domains_with_route53_hosted_zone_name/template.yaml (hunk 1 above)
Comment (timoschilling) [reply examples/2016-10-31/custom_domains_with_route53_hosted_zone_name/template.yaml]: The line below says `Everything below here is optional` but `IpV6` is already optional so it should be under this comment

Comment (mbarneyjr) [reply examples/2016-10-31/custom_domains_with_route53_hosted_zone_name/template.yaml]: Ahhh, I copied this from the `custom_domains_with_route53` example directory

Comment (praneetap) [reply examples/2016-10-31/custom_domains_with_route53_hosted_zone_name/template.yaml]: @timoschilling Ah, that comment was meant to point the difference between using your own distribution vs apigateway one. Not as "this property is optional".

Comment (mbarneyjr) [reply examples/2016-10-31/custom_domains_with_route53_hosted_zone_name/template.yaml]: I actually don't see `IpV6` in the SAM spec, but from looking at the code behind this, it looks like it is optional

Comment (timoschilling) [reply examples/2016-10-31/custom_domains_with_route53_hosted_zone_name/template.yaml]: Sorry didn't read the full comment, you both are right

File: samtranslator/model/api/api_generator.py (hunk 2 above)
Comment (mbarneyjr) [reply samtranslator/model/api/api_generator.py]: Added

**Comment by mbarneyjr** [comment]
@praneetap I added the test cases and fixed a bug in the LogicalID generation thatI found because of it, thanks for the guidance!

**Comment by praneetap** [comment]
@mbarneyjr please run make black, it will auto format for you. That should fix the build.

**Comment by keetonian** [comment]
Looks like it's just a black formatting error. I submitted #1419 to update the black version that our tests use, this PR should be good to go once that is merged. I don't see any other test issues.

**Comment by keetonian** [comment]
Closed/re-opened PR to kick off a new travis run. The black formatting issues should now be resolved.

**Comment by keetonian** [comment]
@mbarneyjr Thank you for this contribution!

**Pull request merged**